src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp

```diff
@@ -799,6 +804,765 @@ CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_GetExternalIPv6
+            *buffer    = static_cast<uint8_t *>(chip::Platform::MemoryAlloc(ReadLength));
+            VerifyOrExit(*buffer != NULL, err = CHIP_ERROR_NO_MEMORY);
+
+            otOperationalDataset activeDataset;
+            otError otErr = otDatasetGetActive(mOTInst, &activeDataset);
+            VerifyOrExit(otErr == OT_ERROR_NONE, err = MapOpenThreadError(otErr));
+            Encoding::LittleEndian::Put64(*buffer, activeDataset.mPendingTimestamp);
+        }
+    }
+    break;
+
+    case ZCL_DELAY_ATTRIBUTE_ID: {
+        type = kTLVType_UnsignedInteger;
+        if (otDatasetIsCommissioned(mOTInst))
+        {
+            ReadLength = sizeof(uint32_t);
+            *buffer    = static_cast<uint8_t *>(chip::Platform::MemoryAlloc(ReadLength));
+            VerifyOrExit(*buffer != NULL, err = CHIP_ERROR_NO_MEMORY);
+
+            otOperationalDataset activeDataset;
+            otError otErr = otDatasetGetActive(mOTInst, &activeDataset);
+            VerifyOrExit(otErr == OT_ERROR_NONE, err = MapOpenThreadError(otErr));
+            Encoding::LittleEndian::Put32(*buffer, activeDataset.mDelay);
+        }
+    }
+    break;
+
+    case ZCL_SECURITY_POLICY_ATTRIBUTE_ID: {
+        type = kTLVType_Structure;
+        if (otDatasetIsCommissioned(mOTInst))
+        {
+            ReadLength = sizeof(otSecurityPolicy);
+            *buffer    = static_cast<uint8_t *>(chip::Platform::MemoryAlloc(ReadLength));
+            VerifyOrExit(*buffer != NULL, err = CHIP_ERROR_NO_MEMORY);
+
+            otOperationalDataset activeDataset;
+            otError otErr = otDatasetGetActive(mOTInst, &activeDataset);
+            VerifyOrExit(otErr == OT_ERROR_NONE, err = MapOpenThreadError(otErr));
+            memcpy(*buffer, &activeDataset.mSecurityPolicy, sizeof(otSecurityPolicy));
+        }
+    }
+    break;
+
+    case ZCL_DIAG_CHANNEL_MASK_ATTRIBUTE_ID: {
+        type = kTLVType_ByteString;
+        if (otDatasetIsCommissioned(mOTInst))
+        {
+            ReadLength = (sizeof(uint32_t) * 2) + 1; // For all Hex characters + NULL
+            *buffer    = static_cast<uint8_t *>(chip::Platform::MemoryAlloc(ReadLength));
+            VerifyOrExit(*buffer != NULL, err = CHIP_ERROR_NO_MEMORY);
+
+            otOperationalDataset activeDataset;
+            otError otErr = otDatasetGetActive(mOTInst, &activeDataset);
+            VerifyOrExit(otErr == OT_ERROR_NONE, err = MapOpenThreadError(otErr));
+
+            // In the resultant Octet string, the most significant bit of the left-most byte indicates channel 0
+            uint32_t swapedChannelMask = Encoding::Swap32(activeDataset.mChannelMask);
+            chip::Encoding::BytesToUppercaseHexString(reinterpret_cast<uint8_t *>(&swapedChannelMask), sizeof(uint32_t),
+                                                      reinterpret_cast<char *>(*buffer), ReadLength);
+        }
+    }
+    break;
+
+    case ZCL_OPERATIONAL_DATASET_COMPONENTS_ATTRIBUTE_ID: {
+        type = kTLVType_Structure;
```

Comment:
This is not right....  This needs to actually encode a TLV struct, right?

Reply:
How is this done? I wante copy the memory content of activeDataset.mComponents to the allocate buffer for the caller to then encode TLV struct. Any reference on how TLV struct will be encode for attributes?

Reply:
The caller can do the TLV encoding, sure, as long as it knows what it's working with.  If there's only one structure involved, the type might be enough to tell it that this is that structure....

(Right now the caller just errors out on `type == kTLVType_Structure`, so this code is not really exercised.  Once we land codegen for encoding spec structs to TLV, which @yunhanw-google is working on, the caller would need to build the spec struct, then call `Encode` on it.)